### PR TITLE
Parallel image encoding

### PR DESCRIPTION
**Pull request by Dherse**
This makes image encoding use a thread pool iif we're exporting to PDF (hence the modification of `World`). This allows PNG heavy documents to be much faster, in the case of my thesis (as an example) it shaves 38% of the compile time on average.

I am also tempted to do the same for encoding of SVGs to `Chunk`, but I'll wait for @laurmaedje to give me his opinion.

### CODE REVIEW

**Comment by Enter-tainer** [comment]
adding is_pdf everywhere looks a bit of hacky for me. it might be better to change it to `enable_parallel_png_encode` or something.

File: crates/typst/src/image/raster.rs [hunk 1]
```diff
@@ -137,3 +155,65 @@ fn format_image_error(error: image::ImageError) -> EcoString {
...
+        return None;
+    }
+
+    if format == RasterFormat::Jpg && matches!(&**dynamic, DynamicImage::ImageRgb8(_)) {
+        return None;
+    }
+
+    // Special case for encoding luma images.
+    let out = Arc::new(OnceCell::new());
+    if format == RasterFormat::Png && matches!(&**dynamic, DynamicImage::ImageLuma8(_)) {
```
Comment (Enter-tainer) [inline crates/typst/src/image/raster.rs:181]: what is &**? is it possible to use take the inner data out of ImageLuma8 and avoid the unreachable!() in below?

File: crates/typst/src/image/raster.rs [hunk 2]
```diff
@@ -47,7 +58,9 @@ impl RasterImage {
         }
         .map_err(format_image_error)?;
 
-        Ok(Self(Arc::new(Repr { data, format, dynamic, icc })))
+        let dynamic = Arc::new(dynamic);
+        let encoded = encode_image(&dynamic, is_pdf, format);
```
Comment (Enter-tainer) [inline crates/typst/src/image/raster.rs:62]: images are encoded once it is created, and we rely on this to get concurrency

File: crates/typst/src/image/mod.rs [hunk 3]
```diff
@@ -24,14 +25,21 @@ use crate::World;
...
     kind: ImageKind,
     /// A text describing the image.
     alt: Option<EcoString>,
 }
 
+impl Hash for Repr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        // We do not hash the encoded value because it is irrelevant.
+        self.kind.hash(state);
+        self.alt.hash(state);
```
Comment (Enter-tainer) [inline crates/typst/src/image/mod.rs:39]: is this hash impl here same as the derived one? it seems that it simply hashes every fields

File: crates/typst/src/image/raster.rs [hunk 4]
```diff
@@ -7,12 +7,22 @@ use image::codecs::gif::GifDecoder;
...
+};
+use once_cell::sync::{Lazy, OnceCell};
 use typst_macros::Cast;
 
 use crate::diag::{bail, StrResult};
 use crate::eval::Bytes;
 
+static THREAD_POOL: Lazy<rayon::ThreadPool> = Lazy::new(|| {
+    rayon::ThreadPoolBuilder::new()
+        .thread_name(|i| format!("typst-image-encode-{i}"))
```
Comment (Enter-tainer) [inline crates/typst/src/image/raster.rs:21]: i guess the thread pool here will be used for all compute heavy tasks in future.

**Comment by Dherse** [comment]
After another quick discussion on Discord, we found a very neat way of doing it (although still reliant on the `OnceCell` approach) that is still localized to PDF export. It works great and solves all of the aforementioned issues! 🎉

File: crates/typst/src/image/raster.rs (hunk 2 above)
Comment (Dherse) [reply crates/typst/src/image/raster.rs]: Yes, and the `OnceCell` will store the result once the computation is done. Now it's done directly in typst-pdf and it actually shows that we could do this for a lot more things 😈

File: crates/typst/src/image/raster.rs (hunk 4 above)
Comment (Dherse) [reply crates/typst/src/image/raster.rs]: I am now using the global one which solves this issue.

**Comment by laurmaedje** [comment]
I've tested with and without yield_now in the web app and the former indeed deadlocks whereas the latter works. This in line with the description here: https://docs.rs/rayon-core/latest/rayon_core/#global-fallback-when-threading-is-unsupported

**Comment by Dherse** [comment]
At least we have a workaround, thanks for testing

**Comment by Dherse** [comment]
I have moved the deferred logic to its own module in `typst::util` which makes it less obscure to read and self contained (including the rayon workaround) as well as easier to re-use.

**Comment by Dherse** [comment]
BTW, the impl of `Clone` etc. are manual because the default behaviour is to require `B` to be `Clone` as well when derived which just doesn't need to be the case since it's an `Arc<B>`.

**Comment by laurmaedje** [comment]
From the looks of it `Deferred` only holds the initial value to be used in the PDF Remapper construct. I think it's not really it's job though to be a wrapper that retains an initial value, it should just be a deferred computation. By virtue of the deferred image being the remapper key, we also need to lean on comemo to cache `deferred_image` when the same image is used multiple times in the same export process.

Maybe we should keep the Remapper on a normal `Image` and add a separate `encoded_images: HashMap<usize, Deferred<PreEncoded>>` that maps from the PDF indices assigned by the remapper to pre-encoded images. It would be filled via `.entry(idx).get_or_insert_with(|| deferred_image(image.clone()))` so no extra burden on comemo, too.

**Comment by laurmaedje** [comment]
PreEncoded could also be renamed to EncodedImage since it's conceptually separate from the deferral.

**Comment by Dherse** [comment]
I should mention that this does mean we're also encoding SVGs and any future image format supported and this leads (on my thesis) on the latest version to an over 40% reduction in cold compilation time and a ~20% in hot compilation time when modifying images.

**Comment by laurmaedje** [comment]
Thank you!

**Pull request merged**